doc/developer/design/20200520_exactly_once_kafka.md

```diff
@@ -1,28 +1,79 @@
+
+Since all the source operators are also invoked by the same worker, only one thing can happen at a time and there are no race conditions to worry about.
+
+Concretely, every time a source operator is invoked, it needs to:
+ - Check the worker's timestamp bindings for the next potential offset that data might appear at, so that the operator can respect the timestamp assignments of its peers
+ - Downgrade its capability based on the lowest available timestamp across all of the partitions that operator has been assigned to. If the operator has not been assigned to any timestamps it still needs to downgrade its capability to the `current_timestamp`. We cannot close the source operator in this case because the worker may be assigned to a new partition in the future.
+ - Read from its assigned partitions if any, and assign incoming records the appropriate timestamps. If an appropriate binding does not exist the operator needs to extend the timestamp binding for the current timestamp.
+- If the operator minted new timestamp bindings it needs to commit them back to worker-local state before exiting, so that those bindings can be shared with its peers.
+
+Whew. This is a complex (at least for me) protocol that involves some performance changes (now every time we need to timestamp a record we may need to binary search a list of timestamp bindings) to source ingestion code. The only alternative approach which made sense to me for this was to be a Kafka read replica, and write down `(data, timestamp)` for all incoming data, and have only one reader read the topic, and have every other reader read the written down state. This approach was scrapped because of the dependence on source persistence, which is another large project and currently a work in progress.
+
+Once we do the above, we also need a mechanism to compact the list of timestamp bindings in memory, as otherwise we risk running out of memory over time. We can do that rebinding all intervals assigned to a timestamp <= `compaction_frontier` to the `compaction_frontier`. At this stage, we also changed the dataflow worker to communicate source timestamp binding `upper` information back to the `Coordinator` so that the `Coordinator` can direct the dataflow worker on what frontier to compact up to based on the `logical_compaction_window`. After that, the `Coordinator` needs to also be modified to be aware of each source's `upper` and `since` frontiers and to disallow queries outside of the interval `[since, upper)` the way it does for indexes.
```

Comment:
Nit, but strictly speaking the coordinator only disallows queries outside the interval `[since, ..)`. It prefers to answers things within `[.., upper)` but it is secondary to the first constraint.

Reply:
fixed, ty